File: csharp/Zhuxian.Core/Skills.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Zhuxian.Data;

namespace Zhuxian.Core
{
    /// <summary>
    /// 技能与法宝的释放、弹道、命中结算。移植自 src/skills.js。
    ///
    /// 所有技能最终都落到 ApplyHit 上——命中判定各不相同，伤害口径只有一个。
    ///
    /// 释放失败时返回一句人话（灵力不够 / 还在冷却），由 UI 直接显示，
    /// 不静默失败——玩家按了没反应是最糟的手感。
    /// </summary>
    public static class Skills
    {
        // ── 命中结算 ──────────────────────────────────────────

        static void Visual(GameContext game, VisualFx v)
        {
            v.life = v.ms;
            game.Fx.Visuals.Add(v);
        }

        /// <summary>
        /// 单次命中结算：算伤害 → 扣血 → 吸血 → 附加状态 → 飘字。
        /// 这是玩家打怪的唯一入口。
        /// </summary>
        public static double ApplyHit(GameContext game, MonsterInstance m, HitOptions opt)
        {
            var p = game.Player;
            var s = p.stats;
            var isMagic = opt.Dmg == "magic";
            var atkVal = isMagic ? s.mag : s.atk;
            var defVal = isMagic ? (double)m.def.mdef : m.def.def;

            var res = Combat.Damage(atkVal, defVal, opt.Dmg, opt.Power,
                new Combat.DamageOptions
                {
                    CritRate = s.crit,
                    LevelGap = p.level - m.def.lv,
                },
                game.Data.Config, game.Rng);

            var real = game.World.HurtMonster(m, res.Amount, game.Hooks);
            game.Hooks.Floater(m.x, m.y - m.def.radius - 8, real, res.Crit ? "crit" : "hit");

            // 命中点炸一下。伤害数字说的是"扣了多少"，火花说的是"这一下打上了"——
            // 两者缺一个，打击感就差一截
            Visual(game, new VisualFx
            {
                kind = "impact", x = m.x, y = m.y,
                r = res.Crit ? 26 : 17,
                ms = res.Crit ? 260 : 180,
                color = res.Crit ? "#ffd24a" : "#ffffff",
            });

            // 吸血：技能自带的 + 装备/增益的，取和
            var steal = opt.Lifesteal + s.lifesteal;
            if (steal > 0 && !p.dead)
            {
                var back = Combat.Lifesteal(steal, real);
                var room = p.stats.hp - p.hp;
                if (room > 0)
                {
                    var got = Math.Min(room, back);
                    p.hp += got;
                    game.Hooks.Floater(p.x, p.y - 34, got, "heal");
                }
            }

            // 附加状态
            if (opt.Burn != 0)
            {
                Combat.AddBuff(m.buffs, new Buff
                {
                    kind = "burn",
                    ms = opt.BurnMs != 0 ? opt.BurnMs : 4000,
                    dps = atkVal * opt.Burn,
                });
            }
            if (opt.Poison != 0)
            {
                Combat.AddBuff(m.buffs, new Buff
                {
                    kind = "poison",
                    ms = opt.PoisonMs != 0 ? opt.PoisonMs : 6000,
                    dps = atkVal * opt.Poison,
                });
            }
            if (opt.StunMs != 0) Combat.AddBuff(m.buffs, new Buff { kind = "stun", ms = opt.StunMs });
            if (opt.CharmMs != 0) Combat.AddBuff(m.buffs, new Buff { kind = "charm", ms = opt.CharmMs });
            if (opt.Slow != 0)
            {
                Combat.AddBuff(m.buffs, new Buff
                {
                    kind = "slow",
                    ms = opt.SlowMs != 0 ? opt.SlowMs : 3000,
                    amount = opt.Slow,
                });
            }
            if (opt.Knock != 0)
            {
                var d = Geometry.DirTo(p.x, p.y, m.x, m.y);
                var np = game.World.Move(m.x, m.y, d.X * opt.Knock, d.Y * opt.Knock, m.def.radius);
                m.x = np.X;
                m.y = np.Y;
            }

            return real;
        }

        // ── 各种命中形状 ──────────────────────────────────────

        static int DoNova(GameContext game, double x, double y, double radius, HitOptions opt)
        {
            var list = game.World.InRadius(x, y, radius);
            foreach (var m in list) ApplyHit(game, m, opt);

            Visual(game, new VisualFx { kind = "nova", x = x, y = y, r = radius, ms = 380, color = opt.Color });
            game.Hooks.Sfx("nova");
            return list.Count;
        }

        static int DoStrike(GameContext game, HitOptions opt)
        {
            var p = game.Player;
            var range = opt.Range != 0 ? opt.Range : 90;

            // 朝向前方半圆内、射程内的怪，按距离取最多 3 个
            var cands = game.World.InRadius(p.x, p.y, range);
            var hits = new List<MonsterInstance>();
            foreach (var m in cands)
            {
                var d = Geometry.DirTo(p.x, p.y, m.x, m.y);
                if (d.X * p.facingX + d.Y * p.facingY > -0.1) hits.Add(m);
            }

            // 用 OrderBy 而不是 List.Sort：OrderBy 是稳定排序，和 JS 的 Array.sort 一致。
            // List.Sort 不稳定，距离相同的两只怪谁在前是不确定的——多段技能打谁不打谁就会飘
            hits = hits
                .OrderBy(m => Geometry.Dist2(p.x, p.y, m.x, m.y))
                .Take(3)
                .ToList();

            // 让人物做出挥砍动作。多段技能也只播一次，不然会抽搐
            p.swingMs = game.Data.Config.SWING_MS;

            var times = opt.Hits != 0 ? opt.Hits : 1;
            for (var t = 0; t < times; t++)
            {
                foreach (var m in hits)
                {
                    if (!m.dead) ApplyHit(game, m, opt);
                }
            }

            Visual(game, new VisualFx
            {
                kind = "slash", x = p.x, y = p.y, r = range,
                ax = p.facingX, ay = p.facingY, ms = 220, color = opt.Color,
            });
            game.Hooks.Sfx("slash");
            return hits.Count;
        }

        static void DoPierce(GameContext game, double tx, double ty, HitOptions opt)
        {
            var p = game.Player;
            var dir = Geometry.DirTo(p.x, p.y, tx, ty);
            game.Fx.Bolts.Add(new Bolt
            {
                uid = Uid.Next(),
                kind = "pierce",
                x = p.x, y = p.y,
                dx = dir.X, dy = dir.Y,
                speed = opt.Speed != 0 ? opt.Speed : 600,
                left = opt.Range != 0 ? opt.Range : 360,
                width = opt.Width != 0 ? opt.Width : 26,
                opt = opt,
                color = opt.Color,
            });
            game.Hooks.Sfx("cast");
        }

        static void DoBolt(GameContext game, double tx, double ty, HitOptions opt)
        {
            var p = game.Player;
            var dir = Geometry.DirTo(p.x, p.y, tx, ty);
            game.Fx.Bolts.Add(new Bolt
            {
                uid = Uid.Next(),
                kind = "bolt",
                x = p.x, y = p.y,
                dx = dir.X, dy = dir.Y,
                speed = opt.Speed != 0 ? opt.Speed : 420,
                left = opt.Range != 0 ? opt.Range : 320,
                width = 14,
                opt = opt,
                color = opt.Color,
            });
            game.Hooks.Sfx("cast");
        }

        static void DoBlast(GameContext game, double tx, double ty, HitOptions opt)
        {
            var p = game.Player;

            // 落点钳在射程内，点太远就打在射程边上，而不是放空
            var d = Geometry.Dist(p.x, p.y, tx, ty);
            var max = opt.Range != 0 ? opt.Range : 300;
            if (d > max)
            {
                var dir = Geometry.DirTo(p.x, p.y, tx, ty);
                tx = p.x + dir.X * max;
                ty = p.y + dir.Y * max;
            }

            var dir2 = Geometry.DirTo(p.x, p.y, tx, ty);
            game.Fx.Bolts.Add(new Bolt
            {
                uid = Uid.Next(),
                kind = "blast",
                x = p.x, y = p.y,
                tx = tx, ty = ty,
                dx = dir2.X, dy = dir2.Y,
                speed = 560,
                left = Geometry.Dist(p.x, p.y, tx, ty) + 1,
                width = 12,
                opt = opt,
                color = opt.Color,
            });
            game.Hooks.Sfx("cast");
        }

        // ── 释放 ──────────────────────────────────────────────

        /// <summary>技能定义摊平成命中参数</summary>
        static HitOptions OptionsOf(ZxSectSkill def)
        {
            return new HitOptions
            {
                Dmg = def.dmg,
                Power = def.power,
                Lifesteal = def.lifesteal,
                Burn = def.burn,
                BurnMs = def.burnMs,
                StunMs = def.stunMs,
                Slow = def.slow,
                SlowMs = def.slowMs,
                Knock = def.knock,
                Hits = def.hits,
                Range = def.range,
                Speed = def.speed,
                Width = def.width,
                Radius = def.radius,
                Color = def.color,
            };
        }

        /// <summary>
        /// 放技能。tx/ty 是瞄准点（鼠标位置，或摇杆方向上的一点）。
        /// </summary>
        public static CastResult Cast(GameContext game, ZxSectSkill def, double tx, double ty)
        {
            var p = game.Player;
            if (p.dead) return CastResult.Fail("你已经倒下了。");
            if (Combat.Disabled(p.buffs)) return CastResult.Fail("动弹不得！");

            var cd = p.cd.TryGetValue(def.key, out var left) ? left : 0;
            if (cd > 0)
            {
                return CastResult.Fail(def.name + " 还需 " + (cd / 1000).ToString("F1") + " 秒");
            }
            if (p.mp < def.mp) return CastResult.Fail("灵力不足。");

            var costHp = def.costHp != 0 ? Math.Floor(p.stats.hp * def.costHp) : 0;
            if (costHp != 0 && p.hp <= costHp) return CastResult.Fail("气血不足以支撑血炼。");

            p.mp -= def.mp;
            if (costHp != 0)
            {
                p.hp -= costHp;
                game.Hooks.Floater(p.x, p.y - 34, costHp, "hurt");
            }
            p.cd[def.key] = def.cd;

            var opt = OptionsOf(def);

            switch (def.kind)
            {
                case "strike": DoStrike(game, opt); break;
                case "bolt": DoBolt(game, tx, ty, opt); break;
                case "pierce": DoPierce(game, tx, ty, opt); break;
                case "nova": DoNova(game, p.x, p.y, def.radius, opt); break;
                case "blast": DoBlast(game, tx, ty, opt); break;
                case "buff":
                case "heal":
                    // 这两类没有命中形状，效果全在下面的自身增益部分
                    break;
            }

            // 自身增益部分。很多攻击技也带护盾，所以独立于 kind 判断
            if (def.shield != 0)
            {
                Combat.AddBuff(p.buffs, new Buff
                {
                    kind = "shield",
                    ms = def.shieldMs != 0 ? def.shieldMs : 8000,
                    value = Math.Floor(p.stats.hp * def.shield),
                });
                Visual(game, new VisualFx { kind = "ring", x = p.x, y = p.y, r = 44, ms = 500, color = "#9fd8ff", follow = true });
            }
            if (def.buffAtk != 0)
            {
                Combat.AddBuff(p.buffs, new Buff
                {
                    kind = "atkUp",
                    ms = def.buffMs != 0 ? def.buffMs : 8000,
                    amount = def.buffAtk,
                });
                Player.Recompute(game.Data, p);
            }
            if (def.buffDef != 0)
            {
                Combat.AddBuff(p.buffs, new Buff
                {
                    kind = "defUp",
                    ms = def.buffMs != 0 ? def.buffMs : 8000,
                    amount = def.buffDef,
                });
                Player.Recompute(game.Data, p);
            }
            if (def.heal != 0)
            {
                var amount = Math.Floor(p.stats.hp * def.heal) + def.healFlat;
                var got = Math.Min(amount, p.stats.hp - p.hp);
                p.hp += got;
                game.Hooks.Floater(p.x, p.y - 34, got, "heal");
                Visual(game, new VisualFx { kind = "ring", x = p.x, y = p.y, r = 40, ms = 520, color = "#8ef0b0", follow = true });
                game.Hooks.Sfx("heal");
            }

            game.Hooks.OnCast(def);
            return CastResult.Success;
        }

        /// <summary>法宝主动技。和门派技能共用冷却表，key 用 talisman</summary>
        public static CastResult CastTalisman(GameContext game, double tx, double ty)
        {
            var p = game.Player;
            var item = Inventory.Talisman(game.Data, p.equip);
            if (item == null || item.active == null || !item.active.present)
            {
                return CastResult.Fail("没有装备法宝。");
            }
            if (p.dead) return CastResult.Fail("你已经倒下了。");

            var a = item.active;
            var cd = p.cd.TryGetValue("talisman", out var left) ? left : 0;
            if (cd > 0)
            {
                return CastResult.Fail(item.name + " 尚未凝聚（" + (cd / 1000).ToString("F1") + " 秒）");
            }
            p.cd["talisman"] = a.cd;

            switch (a.kind)
            {
                case "heal":
                {
                    var amount = Math.Floor(p.stats.hp * a.heal) + a.healFlat;
                    var got = Math.Min(amount, p.stats.hp - p.hp);
                    p.hp += got;
                    game.Hooks.Floater(p.x, p.y - 34, got, "heal");
                    Visual(game, new VisualFx { kind = "ring", x = p.x, y = p.y, r = 46, ms = 600, color = "#8ef0b0", follow = true });
                    game.Hooks.Sfx("heal");
                    break;
                }
                case "nova":
                    DoNova(game, p.x, p.y, a.radius, new HitOptions
                    {
                        Dmg = "magic", Power = a.power, Burn = a.burn, BurnMs = a.burnMs, Color = "#ff9a4a",
                    });
                    break;
                case "stun":
                {
                    foreach (var m in game.World.InRadius(p.x, p.y, a.radius))
                    {
                        Combat.AddBuff(m.buffs, new Buff { kind = "stun", ms = a.ms });
                    }
                    Visual(game, new VisualFx { kind = "nova", x = p.x, y = p.y, r = a.radius, ms = 420, color = "#b48ff0" });
                    game.Hooks.Sfx("nova");
                    break;
                }
                case "charm":
                {
                    foreach (var m in game.World.InRadius(p.x, p.y, a.radius))
                    {
                        Combat.AddBuff(m.buffs, new Buff { kind = "charm", ms = a.ms });
                    }
                    Visual(game, new VisualFx { kind = "nova", x = p.x, y = p.y, r = a.radius, ms = 420, color = "#f0a0c8" });
                    game.Hooks.Sfx("bell");
                    break;
                }
                case "poison":
                {
                    foreach (var m in game.World.InRadius(p.x, p.y, a.radius))
                    {
                        Combat.AddBuff(m.buffs, new Buff { kind = "poison", ms = a.ms, dps = p.stats.atk * a.power });
                    }
                    Visual(game, new VisualFx { kind = "nova", x = p.x, y = p.y, r = a.radius, ms = 520, color = "#8fc44a" });
                    game.Hooks.Sfx("nova");
                    break;
                }
                case "lifesteal":
                    Combat.AddBuff(p.buffs, new Buff { kind = "lifesteal", ms = a.ms, amount = a.amount });
                    Player.Recompute(game.Data, p);
                    Visual(game, new VisualFx { kind = "ring", x = p.x, y = p.y, r = 42, ms = 600, color = "#e05a6e", follow = true });
                    break;
                case "reflect":
                    Combat.AddBuff(p.buffs, new Buff { kind = "reflect", ms = a.ms, amount = a.ratio });
                    Visual(game, new VisualFx { kind = "ring", x = p.x, y = p.y, r = 46, ms = 600, color = "#9ae0ff", follow = true });
                    break;
                case "shield":
                    Combat.AddBuff(p.buffs, new Buff { kind = "shield", ms = a.ms, value = Math.Floor(p.stats.hp * a.shield) });
                    Visual(game, new VisualFx { kind = "ring", x = p.x, y = p.y, r = 48, ms = 600, color = "#ffd98a", follow = true });
                    break;
            }

            game.Hooks.Log("祭出【" + item.name + "】· " + a.name, "skill");
            return CastResult.Success;
        }

        /// <summary>普通攻击：近身扇形，走和技能一样的结算</summary>
        public static bool BasicAttack(GameContext game)
        {
            var p = game.Player;
            var cfg = game.Data.Config;
            var target = game.World.Nearest(p.x, p.y, cfg.MELEE_RANGE + 20);
            if (target == null) return false;

            var dir = Geometry.DirTo(p.x, p.y, target.x, target.y);
            p.facingX = dir.X;
            p.facingY = dir.Y;

            DoStrike(game, new HitOptions
            {
                Dmg = "phys",
                Power = 1,
                Range = cfg.MELEE_RANGE,
                Color = game.Data.Sect(p.sect).color,
            });
            return true;
        }

        // ── 每帧推进 ──────────────────────────────────────────

        public static void Update(GameContext game, double dt)
        {
            var p = game.Player;

            // 冷却
            foreach (var key in p.cd.Keys.ToList())
            {
                if (p.cd[key] > 0) p.cd[key] = Math.Max(0, p.cd[key] - dt);
            }

            // 飞行物
            for (var i = game.Fx.Bolts.Count - 1; i >= 0; i--)
            {
                var b = game.Fx.Bolts[i];
                var step = b.speed * (dt / 1000.0);
                b.x += b.dx * step;
                b.y += b.dy * step;
                b.left -= step;

                var done = false;

                // 撞墙就停（blast 停在墙前也照样炸）
                if (game.World.Blocked(b.x, b.y, 4)) done = true;

                foreach (var m in game.World.InRadius(b.x, b.y, b.width))
                {
                    if (b.hit.Contains(m.uid)) continue;
                    b.hit.Add(m.uid);
                    if (b.kind == "blast") continue; // 落点炸，路上不碰

                    ApplyHit(game, m, b.opt);
                    if (b.kind == "bolt")
                    {
                        done = true;
                        break;
                    }
                }

                if (b.left <= 0) done = true;

                if (!done) continue;

                if (b.kind == "blast")
                {
                    DoNova(game, b.x, b.y, b.opt.Radius != 0 ? b.opt.Radius : 120, b.opt);
                }
                else
                {
                    Visual(game, new VisualFx { kind = "spark", x = b.x, y = b.y, r = 16, ms = 200, color = b.color });
                }
                game.Fx.Bolts.RemoveAt(i);
            }

            // 特效寿命
            for (var i = game.Fx.Visuals.Count - 1; i >= 0; i--)
            {
                var v = game.Fx.Visuals[i];
                v.life -= dt;
                if (v.follow)
                {
                    v.x = p.x;
                    v.y = p.y;
                }
                if (v.life <= 0) game.Fx.Visuals.RemoveAt(i);
            }
        }
    }
}

File: tools/fixtures/skills.mjs
/**
 * 对照样本的内容定义。由 tools/gen-fixtures.mjs 调用。
 *
 * 每个 build* 函数拿到一个上下文：
 *   ZX          跑起来的 JS 游戏模块
 *   mulberry32  确定性随机，和 C# 侧 Zhuxian.Core/Rng.cs 是同一个算法
 *   withRandom  在一段代码里接管 Math.random，并记下这段用掉了哪些随机数
 *
 * 挑样本的原则：挑那些「翻错了但看着还挺对」的地方——取整方向、边界条件、
 * 随机数消耗顺序、以及"背包满了会怎样"这类分支。纯直给的公式反而不用堆太多条。
 */

// 技能释放、弹道、法宝主动。对应 src/skills.js。
/**
 * 技能释放与弹道。这块是随机数消耗最密集的地方：
 * 一发范围技能打中 5 只怪就是 5 次伤害结算、最多 10 个随机数，
 * 顺序错一个后面全歪，所以每条用例都记下"这段一共摇了几次"。
 *
 * 覆盖点挑的是翻译最容易走样的地方：
 *   - 释放失败的几种理由（冷却 / 灵力 / 倒下 / 定身）要一句不差
 *   - strike 只打前方半圆内最近的 3 个，而且排序要稳定
 *   - bolt 撞一只就停，pierce 穿过去继续飞，blast 路上不碰、落点才炸
 *   - 血炼类技能扣自己的血；护盾 / 增益要走 recompute
 */
export function buildSkills({ ZX, withRandom }) {
  const CFG = ZX.CONFIG;

  /** 造一局：一张图、一个玩家、一套记录用的钩子 */
  function newGame(seed, sectKey, level) {
    ZX.U.seedUid(0);
    const world = ZX.World.create('caomiao');
    const player = ZX.Player.create('测试', sectKey || 'qingyun');
    if (level) ZX.Player.gainExp(player, 999999999);
    player.level = level || player.level;
    ZX.Player.recompute(player);
    player.hp = player.stats.hp;
    player.mp = player.stats.mp;
    player.x = world.w / 2;
    player.y = world.h / 2;
    player.facing = { x: 1, y: 0 };

    const events = [];
    const game = {
      player: player,
      world: world,
      fx: ZX.Skills.newFx(),
      hooks: {
        floater: (x, y, amount, kind) => events.push({ kind: 'floater', amount: amount, tag: kind }),
        sfx: (name) => events.push({ kind: 'sfx', tag: name }),
        log: (msg, kind) => events.push({ kind: 'log', tag: kind }),
        onCast: (def) => events.push({ kind: 'cast', tag: def.key }),
        onDeath: (m) => events.push({ kind: 'death', tag: m.def.id }),
        onDot: () => {},
        onHitPlayer: () => {},
      },
    };
    return { game: game, events: events, seed: seed };
  }

  /** 把怪摆到玩家跟前，免得靠刷怪的随机位置碰运气 */
  function placeMonsters(game, spec) {
    const p = game.player;
    game.world.monsters.length = 0;
    for (const [id, dx, dy] of spec) {
      const def = ZX.MONSTERS.byId(id);
      const m = {
        uid: ZX.U.uid(), def: def,
        x: p.x + dx, y: p.y + dy, homeX: p.x + dx, homeY: p.y + dy,
        hp: def.maxHp, maxHp: def.maxHp,
        state: 'idle', atkCd: 0, wanderCd: 9999,
        vx: 0, vy: 0, buffs: [], flash: 0, dead: false, castFx: 0,
      };
      game.world.monsters.push(m);
    }
  }

  // 故意不记 uid：uid 是个全局自增计数器，记了的话这组样本就隐式依赖
  // 「前面几组一共用掉多少 uid」，加一条别的用例就会把这里全打乱。
  // 这组怪是手工按顺序摆的，按下标认人就够了。
  const snapMonsters = (game) =>
    game.world.monsters.map((m) => ({
      id: m.def.id, hp: m.hp, x: m.x, y: m.y, dead: !!m.dead,
      buffs: m.buffs.map((b) => ({ kind: b.kind, ms: b.ms, dps: b.dps == null ? 0 : b.dps, amount: b.amount == null ? 0 : b.amount })),
    }));

  const snapBolts = (game) =>
    game.fx.bolts.map((b) => ({ kind: b.kind, x: b.x, y: b.y, dx: b.dx, dy: b.dy, speed: b.speed, left: b.left, width: b.width }));

  const snapPlayer = (p) => ({ hp: p.hp, mp: p.mp, swingMs: p.swingMs, atk: p.stats.atk, def: p.stats.def, buffs: p.buffs.map((b) => b.kind) });

  // 1. 释放失败的几种理由
  const failures = [];
  {
    const { game } = newGame(5000);
    const p = game.player;
    const sect = ZX.sect(p.sect);
    const skill = sect.skills[0];

    // 灵力不足
    p.mp = 0;
    failures.push({ step: '灵力不足', skillKey: skill.key, ok: ZX.Skills.cast(game, skill, p.x + 100, p.y).ok, msg: ZX.Skills.cast(game, skill, p.x + 100, p.y).msg });

    // 冷却中
    p.mp = p.stats.mp;
    ZX.Skills.cast(game, skill, p.x + 100, p.y);
    const cooling = ZX.Skills.cast(game, skill, p.x + 100, p.y);
    failures.push({ step: '冷却中', skillKey: skill.key, ok: cooling.ok, msg: cooling.msg });

    // 定身
    p.cd[skill.key] = 0;
    ZX.Combat.addBuff(p, { kind: 'stun', ms: 2000 });
    const stunned = ZX.Skills.cast(game, skill, p.x + 100, p.y);
    failures.push({ step: '被定身', skillKey: skill.key, ok: stunned.ok, msg: stunned.msg });

    // 倒下
    p.buffs.length = 0;
    p.dead = true;
    const down = ZX.Skills.cast(game, skill, p.x + 100, p.y);
    failures.push({ step: '已倒下', skillKey: skill.key, ok: down.ok, msg: down.msg });
  }

  // 2. 各门派各技能各放一次，怪固定摆位
  const casts = [];
  for (const sect of ZX.SECTS) {
    for (const skill of sect.skills) {
      const seed = 5100 + casts.length;
      const { game, events } = newGame(seed, sect.key, 65);
      placeMonsters(game, [
        ['yegou', 40, 0], ['shanzhu', 70, 30], ['qingshe', -60, 10],
        ['shanzei', 120, -20], ['linghou', 200, 5],
      ]);
      const p = game.player;

      const { value: res, used } = withRandom(seed, () => ZX.Skills.cast(game, skill, p.x + 300, p.y));

      casts.push({
        seed: seed, sect: sect.key, skillKey: skill.key, kind: skill.kind,
        ok: res.ok, msg: res.msg == null ? null : res.msg,
        randomsUsed: used.length,
        player: snapPlayer(p),
        monsters: snapMonsters(game),
        bolts: snapBolts(game),
        visuals: game.fx.visuals.map((v) => ({ kind: v.kind, r: v.r, ms: v.ms })),
        events: events,
      });
    }
  }

  // 3. 弹道推进：bolt 撞一只就停，pierce 穿过去继续飞，blast 落点才炸
  const flights = [];
  for (const [seed, sectKey, skillKey] of [[5300, 'qingyun', 'qy1'], [5301, 'qingyun', 'qy2'], [5302, 'qingyun', 'qy3']]) {
    const { game, events } = newGame(seed, sectKey, 65);
    placeMonsters(game, [['yegou', 80, 0], ['shanzhu', 160, 0], ['qingshe', 240, 0]]);
    const p = game.player;
    const skill = ZX.sect(sectKey).skills.find((s) => s.key === skillKey);

    const { used } = withRandom(seed, () => {
      ZX.Skills.cast(game, skill, p.x + 300, p.y);
      for (let i = 0; i < 20; i++) ZX.Skills.update(game, 50);
      return null;
    });

    flights.push({
      seed: seed, skillKey: skillKey, kind: skill.kind, randomsUsed: used.length,
      monsters: snapMonsters(game), bolts: snapBolts(game),
      visuals: game.fx.visuals.length, events: events,
    });
  }

  // 4. 普攻：面朝最近的怪，最多打 3 个
  const basics = [];
  {
    const seed = 5400;
    const { game, events } = newGame(seed, 'qingyun', 30);
    placeMonsters(game, [
      ['yegou', 30, 0], ['shanzhu', 45, 10], ['qingshe', 20, -20],
      ['shanzei', 35, 25], ['linghou', -300, 0],
    ]);
    const { value: hit, used } = withRandom(seed, () => ZX.Skills.basicAttack(game));
    basics.push({
      seed: seed, hit: hit, randomsUsed: used.length,
      facingX: game.player.facing.x, facingY: game.player.facing.y,
      monsters: snapMonsters(game), events: events,
    });

    // 周围没怪时打空
    const { game: g2 } = newGame(5401, 'qingyun', 30);
    g2.world.monsters.length = 0;
    basics.push({ seed: 5401, hit: ZX.Skills.basicAttack(g2), randomsUsed: 0, facingX: g2.player.facing.x, facingY: g2.player.facing.y, monsters: [], events: [] });
  }

  // 5. 法宝主动
  const talismans = [];
  {
    // 八种主动效果各挑一件，把 castTalisman 的每条分支都走一遍
    const gear = ['t_jinling', 't_shehun', 't_shixuezhu', 't_hehuanling', 't_xuanhuojian', 't_qiweiwugong', 't_liuhejing', 't_tianya'];
    const seeds = gear.map((_, i) => 5500 + i);
    for (let i = 0; i < gear.length; i++) {
      const item = ZX.ITEMS.byId(gear[i]);
      if (!item || !item.active) continue;
      const seed = seeds[i];
      const { game, events } = newGame(seed, 'qingyun', 65);
      placeMonsters(game, [['yegou', 40, 0], ['shanzhu', 70, 30], ['qingshe', -60, 10]]);
      const p = game.player;
      p.equip.talisman = item.id;
      ZX.Player.recompute(p);
      p.hp = Math.floor(p.stats.hp * 0.5);

      const { value: res, used } = withRandom(seed, () => ZX.Skills.castTalisman(game, p.x + 100, p.y));
      talismans.push({
        seed: seed, itemId: item.id, activeKind: item.active.kind,
        ok: res.ok, msg: res.msg == null ? null : res.msg,
        randomsUsed: used.length,
        player: snapPlayer(p), monsters: snapMonsters(game), events: events,
      });
    }

    // 没装法宝
    const { game: g } = newGame(5599, 'qingyun', 10);
    const none = ZX.Skills.castTalisman(g, 0, 0);
    talismans.push({ seed: 5599, itemId: null, activeKind: null, ok: none.ok, msg: none.msg, randomsUsed: 0, player: snapPlayer(g.player), monsters: [], events: [] });
  }

  // 6. 冷却推进与特效寿命
  const upkeep = [];
  {
    const { game } = newGame(5600, 'qingyun', 65);
    placeMonsters(game, [['yegou', 40, 0]]);
    const p = game.player;
    const skill = ZX.sect('qingyun').skills[0];
    withRandom(5600, () => ZX.Skills.cast(game, skill, p.x + 100, p.y));
    for (const dt of [100, 300, 1000, 5000]) {
      ZX.Skills.update(game, dt);
      upkeep.push({ dt: dt, cd: p.cd[skill.key], bolts: game.fx.bolts.length, visuals: game.fx.visuals.length });
    }
  }

  return { failures: failures, casts: casts, flights: flights, basics: basics, talismans: talismans, upkeep: upkeep };
}
